Add 'Integrated Storage vs. Consul' comparison (#13999)
hashicorp/vault: website/content/docs/configuration/storage/index.mdx

For configuration options which also read an environment variable, the
environment variable will take precedence over values in the configuration
file.

> ## Embedded storage vs. external storage

Reviewer: Would you mind adding a link to this page from the storage concepts page to help with discoverability?

Author: Added a call-out note to reference from the concept page. ([commit](https://github.com/hashicorp/vault/pull/13999/commits/7e5b9ec3fd8e7f8afb1c4986e72be10c139953ac))

![callout-note](https://user-images.githubusercontent.com/7660718/153445289-c9911cdd-ac5f-4e62-a8ad-41cc4dd7b518.png)

Reviewer 2: Small suggestion to consider, but I'd recommend making integrated storage the explicit (or at least implicit) recommendation here, and using the link for more details. This allows someone to understand what we recommend quickly without needing to navigate to and read a separate page (unless they want/need more details).

## Supported Storage Backends

For enterprise customers, HashiCorp provides official support for the following storage backends:
- [Vault's integrated storage](link): the recommended backend for most use cases, available in Vault 1.4 or later
- [Consul storage backend](link): leverage the key-value store of Consul, a multi-cloud service networking
   solution including service mesh, service discovery, and network infrastructure automation

Many options are available with community support, including the above. See our [Storage Configuration](link) section for more information.

[ insert your existing "Choosing a storage backend" call-out ]

HashiCorp recommends using Vault's [integrated
storage](/docs/configuration/storage/raft) for most use cases rather than
configuring another system to store Vault data externally. (Integrated storage is

> an **embedded Vault data storage** available in Vault 1.4 or later.)

Reviewer 3: As a reader, this statement makes more sense here.

| Data location                  | The encrypted Vault data is stored on the same host where the Vault server process runs.                                                                                                                 | The encrypted Vault data is stored where the external storage is located. Therefore, the Vault server and the data storage are hosted on physically separate hosts.                                                |
| System requirements            | Avoid "burstable" CPU and storage options. SSDs should be used for the hard drive. <p />See the [Reference Architecture](https://learn.hashicorp.com/tutorials/vault/raft-reference-architecture) guide. | The host machine should be optimized for the performance of the selected storage backend.                                                                                                                          |

> 

Author: Addressed in [commit](https://github.com/hashicorp/vault/pull/13999/commits/ee6605f8dd6cbbfef76a4c28407d268ec891afaf)

> Another factor to consider may be the cost of running an external system to

Reviewer 3: Without much context, this statement feels out of place here.

> store Vault data. However, it is relative to where you run the external storage.

Reviewer: ```suggestion
store Vault data. However, cost is relative to where you run the external storage.
```

### Integrated storage vs. Consul as Vault storage

[HashiCorp Consul](https://www.consul.io/docs/intro) is a comprehensive
multi-cloud service networking solution including service mesh, service

> discovery, and network infrastructure automation. Vault can leverage the

Reviewer 3: Removing the word 'the' as it's not needed here.

Consul's [KV Store](https://www.consul.io/api-docs/kv) to persist Vault data.

> Prior to Vault 1.4, Consul was the recommended Vault storage.

Reviewer 3: ```suggestion
```

Reviewer 3: Suggested to move up above

> Prior to Vault 1.4, Consul was the recommended Vault storage.

Author: @aartiiyengar You requested to have this intro added.  Any thought?

> Prior to Vault 1.4, Consul was the recommended Vault storage.

Reviewer 4: The purpose behind this was to frame Consul in the light of its broader capabilities and lead to the fact that Vault can leverage its KV storage. I think this is good to have. The fact that we recommended IS as the Vault storage has been called out clearly in the doc.

The table below highlights the differences between integrated storage and Consul

> in addition to the table above.

Reviewer 3: Recommend omitting for clarity

| Deployment          | Vault cluster is all you need.                                                                               | Vault cluster & Consul cluster. <p />Use a dedicated Consul cluster for Vault storage, and it should not be used for other purposes (e.g., service discovery, service mesh). <p />See the [Vault with Consul Storage Reference Architecture](https://learn.hashicorp.com/tutorials/vault/reference-architecture#recommended-architecture) guide. |
| Data location       | Data is on disk.                                                                                             | All data is in memory.                                                                                                                                                                                                                                                                                                                           |

> | System requirements | Storage-optimized high IOPS machines.                                                                        | Memory-optimized machines.                                                                                                                                                                                                                                                                                                                       |

Reviewer 3: Since the Reference Architectures should be the source of truth for the system requirements, I would probably link to those here instead. Because when I read 'storage-optimized' or 'memory optimized' instances, that doesn't align with the general compute instances w/ high IOPS storage that are recommended in the Reference Architectures. I would love to get @daveadams thoughts on this too.

Author: Fixed  ([commit](https://github.com/hashicorp/vault/pull/13999/commits/ee6605f8dd6cbbfef76a4c28407d268ec891afaf))

| Snapshots           | Normal data backup strategy of your organization.                                                            | More frequent snapshots are necessary since data is in memory.                                                                                                                                                                                                                                                                                   |
| Max message size    | 1 MiB (Configurable using the [`max_entry_size`](/docs/configuration/storage/raft#max_entry_size) parameter) | 512 KiB (Configurable using the [`kv_max_value_size`](https://www.consul.io/docs/agent/options#kv_max_value_size) parameter)                                                                                                                                                                                                                     |

If you have a Vault cluster using Consul as its storage backend and wish to

> migrate to integrated storage, read to the following tutorials:

Reviewer: ```suggestion
migrate to integrated storage, read the following tutorials:
```

For configuration options which also read an environment variable, the
environment variable will take precedence over values in the configuration
file.

## Embedded storage vs. external storage

HashiCorp recommends using Vault's [integrated
storage](/docs/configuration/storage/raft) for most use cases rather than
configuring another system to store Vault data externally. (Integrated storage is
an **embedded Vault data storage** available in Vault 1.4 or later.)

-> **NOTE:** [HCP Vault](https://cloud.hashicorp.com/products/vault) clusters

> use integrated storage as its storage backend.

Reviewer 5: ```suggestion
use integrated storage as their storage backend.
```

The table below compares the characteristics of integrated storage and external

> storage. If you decide that the additional operational complexity of external

Reviewer 5: ```suggestion
storage. Suppose you decide that the additional operational complexity of external storage is worth it for your use case. In that case, there are several external storage options to choose from (e.g., [Consul](/docs/configuration/storage/consul), [DynamoDB](/docs/configuration/storage/dynamodb), etc.).
```

> storage is worth it for your use case, there are several external storage

Reviewer 5: ```suggestion

```

> options to choose from (e.g., [Consul](/docs/configuration/storage/consul),

Reviewer 5: ```suggestion

```

> [DynamoDB](/docs/configuration/storage/dynamodb), etc.).

Reviewer 5: ```suggestion

```

| Operation                      | Operationally simpler with no additional software installation required.                                                                                                                                 | Must install and configure the external storage environment outside of Vault. For high availability, the external storage should be clustered.                                                                     |
| Networking                     | One less network hop.                                                                                                                                                                                    | Extra network hop between Vault and the external storage system (e.g., Consul cluster).                                                                                                                            |

> | Troubleshooting and monitoring | Integrated storage is a part of Vault; therefore, Vault is the only system you need to monitor and troubleshoot.                                                                                         | The source of failure could be the external storage; therefore, you need to check the health of both Vault and the external storage. This requires the expertise of the chosen storage and properly monitoring it. |

Reviewer 5: ```suggestion
| Troubleshooting and monitoring | Integrated storage is a part of Vault; therefore, Vault is the only system you need to monitor and troubleshoot.                                                                                         | The source of failure could be the external storage; therefore, you need to check the health of both Vault and the external storage. This requires the examination of the chosen storage and properly monitoring it. |
```

| Deployment          | Vault cluster is all you need.                                                                               | Vault cluster & Consul cluster. <p />Use a dedicated Consul cluster for Vault storage, and it should not be used for other purposes (e.g., service discovery, service mesh). <p />See the [Vault with Consul Storage Reference Architecture](https://learn.hashicorp.com/tutorials/vault/reference-architecture#recommended-architecture) guide. |
| Data location       | Data is on disk.                                                                                             | All data is in memory.                                                                                                                                                                                                                                                                                                                           |

> | System requirements | Storage optimized high IOPS machines.                                                                        | Memory optimized machines.                                                                                                                                                                                                                                                                                                                       |

Reviewer 5: ```suggestion
| System requirements | Storage optimized high IOPS machines.                                                                        | Memory-optimized machines.                                                                                                                                                                                                                                                                                                                       |
```
